### PR TITLE
 Delegate cuDNN convolution operation to CuPy

### PR DESCRIPTION
**Pull request by okuta**
Please merge #3781 and cupy/cupy#715 before this PR.

### CODE REVIEW

**Comment by niboshi** [comment]
#3781 has been merged.

**Comment by okuta** [comment]
I rebased.

**Comment by hvy** [comment]
Note, still awaiting https://github.com/cupy/cupy/pull/715.

**Comment by niboshi** [comment]
cupy/cupy#715 has been merged.

**Comment by hvy** [comment]
jenkins, test this please.

File: chainer/backends/cuda.py [hunk 1]
```diff
@@ -535,7 +532,10 @@ def get_max_workspace_size():
         int: The workspace size for cuDNN.
 
     """
-    return _max_workspace_size
+    # To avoid error on no cuDNN environment
+    if cudnn_enabled:
+        return cudnn.get_max_workspace_size()
+    return 0
```
Comment (hvy) [inline chainer/backends/cuda.py]: This seems to change the behavior for non-cuDNN environment. Should we raise warning then this function is called in a such situation, or is the current behavior maybe acceptable?

File: chainer/backends/cuda.py [hunk 2]
```diff
@@ -547,8 +547,9 @@ def set_max_workspace_size(size):
         size: The workspace size for cuDNN.
 
     """
-    global _max_workspace_size
-    _max_workspace_size = size
+    # To avoid error on no cuDNN environment
+    if cudnn_enabled:
+        cudnn.set_max_workspace_size(size)
```
Comment (hvy) [inline chainer/backends/cuda.py]: Ditto.

File: chainer/functions/connection/convolution_2d.py [hunk 3]
```diff
@@ -13,13 +13,7 @@
 
 if cuda.cudnn_enabled:
     cudnn = cuda.cudnn
```
Comment (hvy) [inline chainer/functions/connection/convolution_2d.py]: Maybe we don't have to define this global submodule? It seems to be used only once/twice.

File: chainer/functions/connection/convolution_nd.py [hunk 4]
```diff
@@ -5,21 +5,13 @@
...
 from chainer import configuration
 from chainer import function_node
-from chainer.functions.connection import convolution_2d
 from chainer.utils import conv
 from chainer.utils import conv_nd
 from chainer.utils import type_check
 
 
 if cuda.cudnn_enabled:
     cudnn = cuda.cudnn
```
Comment (hvy) [inline chainer/functions/connection/convolution_nd.py]: Ditto.

File: chainer/functions/connection/deconvolution_2d.py [hunk 5]
```diff
@@ -14,28 +14,7 @@
 
 if cuda.cudnn_enabled:
     cudnn = cuda.cudnn
```
Comment (hvy) [inline chainer/functions/connection/deconvolution_2d.py]: Ditto.

File: chainer/functions/connection/deconvolution_nd.py [hunk 6]
```diff
@@ -6,21 +6,13 @@
...
 from chainer import function_node
 from chainer.functions.connection import convolution_nd
-from chainer.functions.connection import deconvolution_2d
 from chainer.utils import conv
 from chainer.utils import conv_nd
 from chainer.utils import type_check
 
 
 if cuda.cudnn_enabled:
     cudnn = cuda.cudnn
```
Comment (hvy) [inline chainer/functions/connection/deconvolution_nd.py]: Ditto.

**Comment by hvy** [comment]
Added some comments, mostly cosmetics.

**Comment by hvy** [comment]
Do you mind taking a look at the CI failures as well?

File: chainer/backends/cuda.py (hunk 1 above)
Comment (okuta) [reply chainer/backends/cuda.py]: Previous functions did not raise exceptions in non-cuDNN environment.
I agree with the policy to change in the future. However, this PR holds the behavior of the old function.

File: chainer/functions/connection/convolution_2d.py (hunk 3 above)
Comment (okuta) [reply chainer/functions/connection/convolution_2d.py]: Thanks

**Comment by okuta** [comment]
I fixed.

**Comment by hvy** [comment]
jenkins, test this please.

**Comment by hvy** [comment]
jenkins, test this please.

**Comment by hvy** [comment]
LGTM!

**Comment by hvy** [comment]
Removed backport label after offline discussions (https://github.com/cupy/cupy/pull/715 is not going to be backported neither).

**Comment by okuta** [comment]
Thanks!

**Pull request merged**